tests/test_reordering.f90

```diff
@@ -0,0 +1,180 @@
+#else
+    type(omp_backend_t), target :: omp_backend
+    type(allocator_t), target :: omp_allocator
+#endif
+
+    ! Initialise variables and arrays
+    call MPI_Init(ierr)
+    call MPI_Comm_rank(MPI_COMM_WORLD, nrank, ierr)
+    call MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
+
+#ifdef CUDA
+    ierr = cudaGetDeviceCount(ndevs)
+    ierr = cudaSetDevice(mod(nrank, ndevs)) ! round-robin
+    ierr = cudaGetDevice(devnum)
+#endif
+ 
+    globs%nx = 96
+    globs%ny = 96
+    globs%nz = 96
+ 
+    globs%nx_loc = globs%nx/nproc
+    globs%ny_loc = globs%ny/nproc
+    globs%nz_loc = globs%nz/nproc
+ 
+    globs%n_groups_x = globs%ny_loc*globs%nz_loc/SZ
+    globs%n_groups_y = globs%nx_loc*globs%nz_loc/SZ
+    globs%n_groups_z = globs%nx_loc*globs%ny_loc/SZ
+
+    xdirps%n = globs%nx_loc
+    ydirps%n = globs%ny_loc
+    zdirps%n = globs%nz_loc
+
+    xdirps%n_blocks = globs%n_groups_x
+    ydirps%n_blocks = globs%n_groups_y
+    zdirps%n_blocks = globs%n_groups_z
+ 
+   
+#ifdef CUDA
+    cuda_allocator = cuda_allocator_t([SZ, globs%nx_loc, globs%n_groups_x])
+    allocator => cuda_allocator
+    print*, 'CUDA allocator instantiated'
+
+    cuda_backend = cuda_backend_t(globs, allocator)
+    backend => cuda_backend
+    print*, 'CUDA backend instantiated'
+#else
+    omp_allocator = allocator_t([SZ, globs%nx_loc, globs%n_groups_x])
+    allocator => omp_allocator
+    print*, 'OpenMP allocator instantiated'
+
+    omp_backend = omp_backend_t(globs, allocator)
+    backend => omp_backend
+    print*, 'OpenMP backend instantiated'
+#endif
+
+    backend%xdirps => xdirps
+    backend%ydirps => ydirps
+    backend%zdirps => zdirps
+
+    if (nrank == 0) print*, 'Parallel run with', nproc, 'ranks'
+
+   ! Test indexing only
+    do k=1, xdirps%n
+        do j=1, xdirps%n
+            do i=1, xdirps%n
+                call get_index_dir(i_, j_, k_, i, j, k, dir_X, SZ, xdirps%n, xdirps%n, xdirps%n)
+                call get_index_ijk(i__, j__, k__, i_, j_, k_, dir_X, SZ, xdirps%n, xdirps%n, xdirps%n)
+                if (i /= i__ .or. j /= j__ .or. k/= k__) then
+                    print *, "Error in dir X"
+                    allpass = .false.
+                end if
+                call get_index_dir(i_, j_, k_, i, j, k, dir_Y, SZ, xdirps%n, xdirps%n, xdirps%n)
+                call get_index_ijk(i__, j__, k__, i_, j_, k_, dir_Y, SZ, xdirps%n, xdirps%n, xdirps%n)
+                if (i /= i__ .or. j /= j__ .or. k/= k__) then
+                    print *, "Error in dir Y"
+                    allpass = .false.
+                end if
+                call get_index_dir(i_, j_, k_, i, j, k, dir_Z, SZ, xdirps%n, xdirps%n, xdirps%n)
+                call get_index_ijk(i__, j__, k__, i_, j_, k_, dir_Z, SZ, xdirps%n, xdirps%n, xdirps%n)
+                if (i /= i__ .or. j /= j__ .or. k/= k__) then
+                    print *, "Error in dir Z"
+                    allpass = .false.
+                end if
+            end do
+        end do
+    end do
```

Comment:
Extract as a utility subroutine so you can easily pass `xdirps/ydirps/zdirs` and test across all 3, or does this not make sense to do?

Reply:
Yeah, definitely